Refactoring of CAST FORMAT (#7881)

TreeHunter9 opened the pull request:
I look back at my implementation of CAST FORMAT, and I think it's kinda mess, it's trying to do parsing, validation and execution at the same time. In addition, it bloats the cvt.cpp file.   
In this implementation, I moved CAST FORMAT logic to a separate file, and tried to separate parsing, validation and execution steps, so the code will look less messy (I hope so).
Also, I add more FORMAT validation based on SQL standard.

Commit:
Refactoring of CAST FORMAT

Move CAST FORMAT logic to separate file.
Separation of logic into parsing stage and execution stage.
Add more FORMAT validation based on SQL standard.

Commit:
Add more unit tests for incorrect FORMAT

Commit:
Remove incorrect const specifier

asfernandes commented on src/include/firebird/impl/msg/jrd.h:
```diff
@@ -984,3 +984,4 @@ FB_IMPL_MSG(JRD, 981, trailing_part_of_string, -901, "HY", "000", "There is a tr
 FB_IMPL_MSG(JRD, 982, pattern_cant_be_used_without_other_pattern, -901, "HY", "000", "@1 can't be used without @2")
 FB_IMPL_MSG(JRD, 983, pattern_cant_be_used_without_other_pattern_and_vice_versa, -901, "HY", "000", "@1 can't be used without @2 and vice versa")
 FB_IMPL_MSG(JRD, 984, incompatible_format_patterns, -901, "HY", "000", "@1 incompatible with @2")
+FB_IMPL_MSG(JRD, 985, only_one_pattern_can_be_used, -901, "HY", "000", "Can use only one of this patterns @1")
```
one of these?

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+	#define CVT_FORMAT_FLAG(id, format) constexpr Patterns format = 1llu << id - 1;
+	namespace Format
+	{
+		typedef long long unsigned Patterns;
```
Use SINT64 or FB_UINT64.
Or the std variants like uint64_t.

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+		cb->err(Arg::Gds(isc_incompatible_date_format_with_current_date_type) << string(pattern.data(), pattern.length()));
+	}
+
+//---------------------------------------------------------------------------------------------------------------------
```
Please indent comments.

TreeHunter9 replied:
I'm not quiet get it, you mean `// ----`?

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+				continue;
+			}
+
+			while(true)
```
```suggestion
			while (true)
```

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+				if (formatUpper[--tempPos] == '\\')
+				{
+					int backslashCount = 1;
+					while(formatUpper[--tempPos] == '\\')
```
```suggestion
					while (formatUpper[--tempPos] == '\\')
```

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+		constexpr int values[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
+
+		string roman;
+		for (int i = 0; i < 13; i++)
```
Please use `FB_NELEM`.

asfernandes replied:
Or even better, single array with `std::pair` and "for-each".

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+				hours = 12;
+		}
+
+		return { hours, period };
```
```suggestion
		return {hours, period};
```

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+	{
+		string result;
+
+		auto begin = tokens.begin();
```
```suggestion
		const auto begin = tokens.begin();
```

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+		return result;
+	}
+
+//---------------------------------------------------------------------------------------------------------------------
```
Please indent comments.

asfernandes commented on src/common/CvtFormat.cpp:
```diff
@@ -0,0 +1,1678 @@
+	}
+}
+
+string CVT_datetime_to_format_string(const dsc* desc, const string& format, Callbacks* cb)
```
It looks a bit weird that these public functions are declared in `cvt.h` and `CvtFormats.h` is not a header to be included by users of this new "module". I would instead rename current 'CvtFormat.h` to `CvtFormatImpl.h` and declare the public functions in `CvtFormat.h` (with `CVT_format_` prefix).

Commit:
Fixes after review

Minor code changes.
Move public CVT_FORMAT functions  to CvtFormat.h.
Add new error message if same is used pattern twice in string to datetime conversion.

Commit:
Fix indents in comments

asfernandes commented:
New files should be added to the Windows build.

Commit:
Add new files to windows build

Commit:
Remove constexpr specifier from functions that is not real constexpr

Because std::isdigit and std::isalpha are not constexpr.
